### PR TITLE
Improve readability and manageability of deployment configuration for operator

### LINKED ISSUE #14739: Improve readability and manageability of deployment configuration for operator
### Description

I think the configuration of particular properties should be separated from the main `KeycloakDeployment` class as we assume a lot of changes in this stuff, and the class would rapidly grow; not very well manageable.

For KeycloakDeployment class, that could be improved readability and we should keep a certain "convention" in the method ordering(util/helper methods should be positioned below the main/overriden methods).

Also the null-checking conditions in that class are really huge and we could have some solution for that. I'm aware it's a quite comprehensive process to check whether some object in the CR is null or not.

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by mabartos**
Closes #14739

I tried to add an approach to easily customize first-class citizen properties of the CR. I think the configuration should be separated from the main `KeycloakDeployment` class as we assume a lot of changes in this stuff, and the class would rapidly grow; not very well manageable.

I've also slightly refactored that class in order to improve readability and keep a certain "convention" in the method ordering(util/helper methods should be positioned below the main/overriden methods).
I've also refactored the conditions in that class for better readability; I'm aware it's quite comprehensive process to check whether some object in the CR is null or not.

@vmuzikar @Pepo48 @andre-nascimento-rh @pedroigor Could you please check it if you spot anything crucial?

### CODE REVIEW

**Comment by mabartos** [comment]
Added test for the conditions.

File: operator/src/main/java/org/keycloak/utils/DeploymentStatusUtils.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+public class DeploymentStatusUtils {
+
+    /**
+     * Assume the specified first-class citizens are not included in the general server configuration
+     *
+     * @param keycloakDeployment Keycloak deployment
+     * @param status             Status of the deployment
+     * @param firstClassCitizens First-class citizens in the CR
+     */
+    public static void assumeFirstClassCitizens(KeycloakDeployment keycloakDeployment, KeycloakStatusBuilder status, String... firstClassCitizens) {
```
Comment (mabartos) [inline operator/src/main/java/org/keycloak/utils/DeploymentStatusUtils.java:40]: Not used in this PR, but will be used in the following PRs related to the extension of first-class citizen properties of KC CR. I'd keep it here.

File: operator/src/main/java/org/keycloak/utils/DeploymentStatusUtils.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        var configNames = keycloakDeployment.getServerConfigNames();
+
+        if (CollectionUtil.elementsArePresent(configNames, firstClassCitizens)) {
+            status.addWarningMessage("You need to specify these fields as the first-class citizen of the CR: "
```
Comment (mabartos) [inline operator/src/main/java/org/keycloak/utils/DeploymentStatusUtils.java:44]: It'd probably be good to either change the status to error, or when both are specified, then ignore the property in the server config.

**Comment by mabartos** [comment]
It seems there happened some inconsistency during the refactoring. I'll check the failures.

File: operator/src/main/java/org/keycloak/operator/controllers/config/DeploymentConfig.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import io.fabric8.kubernetes.api.model.apps.StatefulSet;
+import org.keycloak.operator.controllers.KeycloakDeployment;
+import org.keycloak.operator.crds.v2alpha1.deployment.KeycloakStatusBuilder;
+
+/**
+ * An interface to represent a configuration of the particular deployment
+ *
+ * @author <a href="mailto:mabartos@redhat.com">Martin Bartos</a>
+ */
+public interface DeploymentConfig {
```
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/controllers/config/DeploymentConfig.java:29]: At first when see this, I really liked this approach, but the I started wondering whether we really need this in the end, what will be the use cases for it. Currently, there's hostname and TLS. The logic for hostname will be simplified as we drop `INSECURE-DISABLE` as part of #14395 to align it with the dist. So there won't be more special logic needed, rest of it is handled by `KeycloakIngress`. In general, we should limit custom logic around config as much as possible and rather implement the changes at the dist side. Most of the fields will be just pass-through without the need for any special logic. And the few selected cases like TLS maybe can stay inside the Deployment?

File: operator/src/main/java/org/keycloak/utils/DeploymentStatusUtils.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+     * Assume the specified first-class citizens are not included in the general server configuration
+     *
+     * @param keycloakDeployment Keycloak deployment
+     * @param status             Status of the deployment
+     * @param firstClassCitizens First-class citizens in the CR
+     */
+    public static void assumeFirstClassCitizens(KeycloakDeployment keycloakDeployment, KeycloakStatusBuilder status, String... firstClassCitizens) {
+        var configNames = keycloakDeployment.getServerConfigNames();
+
+        if (CollectionUtil.elementsArePresent(configNames, firstClassCitizens)) {
```
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/utils/DeploymentStatusUtils.java:43]: 👍

File: operator/src/main/java/org/keycloak/utils/OperatorConditions.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+     *
+     * @param condition supplied condition
+     */
+    public static boolean checkCondition(Supplier<Boolean> condition) {
+        try {
+            return condition.get();
+        } catch (NullPointerException | NoSuchElementException ignore) {
+            return false;
+        }
+    }
```
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/utils/OperatorConditions.java:59]: I like this but maybe it is generic enough to have it somewhere in KC utils?

**Review by vmuzikar — COMMENTED** [review]
@mabartos Thanks for the PR and the initiative! A small refactor is a good idea.

**Comment by mabartos** [comment]
After a discussion with @vmuzikar, we've decided to keep this PR as a Draft and mainly focus on the implementation of particular first-class citizen fields. This PR may be used later.

**Pull request closed without merging**